README.md

```diff
@@ -2,31 +2,512 @@
+this time, we'll have it return the response with the `.andReturn()` method
+call. This will give us a `MvcResult`. From there, we'll grab the content and
+assert that it equals what we expected.
+
+If we run this test, it should also pass!
+
+### Service Integration Tests
+
+We should have now noticed that we still have a potential gap, which is that
+all the tests we've written so far could pass even if the `getFact()` method
+on the `CatFactService` never actually interfaced with the cat fact API.
+
+Let's fix that with an integration test of the service class. We can generate
+this new class just as we did before by right-clicking in the blank space within
+the `CatFactService` class --> Choosing "Generate..." --> Then "Test...". We'll
+call this class `CatFactServiceIntegrationTest`. Once the class has been
+generated, it should look like this:
+
+```java
+package com.example.springtestingdemo.service;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class CatFactServiceIntegrationTest {
+
+    @Test
+    void getFact() {
+    }
+}
+```
+
+Let's rename the test `getFact()` method to something a little more descriptive,
+like "getFactFromExternalApi":
+
+```java
+    @Test
+    void getFactFromExternalApi() {
+    }
+```
+
+Now we'll implement it to call the `CatFactService`, which in turn, will call
+the external cat fact API. This gets tricky since the cat facts are random. We
+cannot test for a specific value back from the service, so we will do two things
+instead:
+
+1. Test that the return value is not null.
+2. Make sure that we don't get the same fact on two consecutive calls, ensuring
+   that the joke is indeed "random".
+
+```java
+package com.example.springtestingdemo.service;
+
+import com.example.springtestingdemo.dto.CatFactDTO;
+import org.junit.jupiter.api.Test;
+import org.springframework.web.client.RestTemplate;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class CatFactServiceIntegrationTest {
+
+    @Test
+    void getFactFromExternalApi() {
+        CatFactService catFactService = new CatFactService(new RestTemplate());
+        CatFactDTO firstCatFact = catFactService.getFact();
+        assertNotNull(firstCatFact);
+        CatFactDTO secondCatFact = catFactService.getFact();
+        assertNotNull(secondCatFact);
+        assertNotEquals(firstCatFact, secondCatFact);
+    }
+}
+```
+
+Note: With a true random service, it's possible that the same return value could
+be returned from 2 consecutive calls. If we wanted to guard against this, we
+could conditionally make a third call if the 2 first calls resulted in the same
+value being returned. The more consecutive calls we make, the less likely they
+are to all return the same value.
+
+If we run the `CatFactServiceIntegrationTest`, we should see it passes.
+
+Note: This is not a "unit" test because it actually lets the real service (not
```

Comment:
Is there much value in this kind of test? 
In many applications, the URL being hit will change depending on environment (LOCAL, TEST, PROD, etc). It could also cost money to hit an endpoint.
It's also possible that the RestTemplate used in the real application will be configured differently (maybe it'll have a root uri or different object mapper or customizer or whatever).

(My previous job didn't do any of this kind of integration testing, so I don't have any experience with it)

Reply:
@developerjaym - this test was in the original lesson, so I left it in here. But that raises a good point. Whenever I tested the "service" it was always mocked up. And if it wasn't mocked up, it was in an acceptance test (Well, what we are calling the acceptance test in this curriculum. Acceptance testing at my old job meant an entire end-to-end test with the UI.)

I am completely okay with removing it since I really don't think it needs to be tested to this level? And the acceptance test will cover it? Would likek @linda-seiter to also weigh in here?

Reply:
I'm good with removing it to simplify things a bit.  This module is covering a lot of challenging concepts.

Reply:
I will go ahead and remove it then. Thanks Linda for weighing in!